localstack/ext/java/src/main/java/cloud/localstack/docker/LocalstackDocker.java

```diff
@@ -0,0 +1,198 @@
+package cloud.localstack.docker;
+
+import cloud.localstack.LocalstackTestRunner;
+import cloud.localstack.ServiceName;
+import cloud.localstack.docker.command.RegexStream;
+import lombok.Getter;
+import lombok.Setter;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Localstack Docker instance
+ *
+ * @author fabianoo
+ */
+public class LocalstackDocker {
+
+    private static final Logger LOG = Logger.getLogger(LocalstackDocker.class.getName());
+
+    private static final String PORT_CONFIG_FILENAME = "/opt/code/localstack/.venv/lib/python2.7/site-packages/localstack_client/config.py";
+
+    private static final Pattern READY_TOKEN = Pattern.compile("Ready\\.");
+
+    //Regular expression used to parse localstack config to determine default ports for services
+    private static final Pattern DEFAULT_PORT_PATTERN = Pattern.compile("'(\\w+)'\\Q: '{proto}://{host}:\\E(\\d+)'");
+    private static final int SERVICE_NAME_GROUP = 1;
+    private static final int PORT_GROUP = 2;
+
+    private Container localStackContainer;
+
+    /**
+     * This is a mapping from service name to internal ports.  In order to use them, the
+     * internal port must be resolved to an external docker port via Container.getExternalPortFor()
+     */
+    private static Map<String, Integer> serviceToPortMap;
+
+    private static boolean started = false;
+
+    @Setter()
+    private String externalHostName = "localhost";
+    @Setter
+    private boolean pullNewImage = true;
+    @Setter
+    private boolean randomizePorts = false;
+    @Setter
+    private Map<String, String> environmentVariables = new HashMap<>();
+
+    @Getter
+    private static LocalstackDocker localstackDocker = new LocalstackDocker();
+
+    private LocalstackDocker() {}
+
+
+    public void startup() {
+        // make sure the local infrastructure is not running, to avoid port conflicts
+        LocalstackTestRunner.teardownInfrastructure();
+
+        // now create the container
+        if (started) {
+            throw new IllegalStateException("A docker instance is already started.");
+        }
+        started = true;
```

Comment:
nitpick: perhaps better set the variable after the container has been initialized (line 73)? If any of the subsequent lines raise an Exception, and the code calling this method uses some sort of retry logic, it wouldn't be able to proceed, since `started` is already set to `true`, when in fact it may not have been started yet. Only a minor issue, though - not critical.